src/bin/cargo-deadlinks.rs

```diff
@@ -78,35 +85,82 @@ fn main() {
 ///
 /// All *.html files under the root directory will be checked.
-fn determine_dir() -> Vec<PathBuf> {
-    let manifest = MetadataCommand::new()
-        .no_deps()
-        .exec()
-        .unwrap_or_else(|err| {
-            println!("error: {}", err);
-            println!("help: if this is not a cargo directory, use `--dir`");
-            ::std::process::exit(1);
-        });
-    let doc = manifest.target_directory.join("doc");
-
-    // originally written with this impressively bad jq query:
-    // `.packages[] |select(.source == null) | .targets[] | select(.kind[] | contains("test") | not) | .name`
-    manifest
-        .packages
-        .into_iter()
-        .filter(|package| package.source.is_none())
-        .map(|package| package.targets)
+fn determine_dir(no_build: bool) -> Vec<PathBuf> {
+    if no_build {
+        log::warn!("--no-build ignores `doc = false` and may have other bugs");
+        let manifest = MetadataCommand::new()
+            .no_deps()
+            .exec()
+            .unwrap_or_else(|err| {
+                println!("error: {}", err);
+                println!("help: if this is not a cargo directory, use `--dir`");
+                process::exit(1);
+            });
+        let doc = manifest.target_directory.join("doc");
+
+        // originally written with this impressively bad jq query:
+        // `.packages[] |select(.source == null) | .targets[] | select(.kind[] | contains("test") | not) | .name`
+        let iter = manifest
+            .packages
+            .into_iter()
+            .filter(|package| package.source.is_none())
+            .map(|package| package.targets)
+            .flatten()
+            .filter(has_docs)
+            .map(move |target| doc.join(target.name.replace('-', "_")));
+        return iter.collect();
+    }
+
+    // Build the documentation, collecting info about the build at the same time.
+    log::info!("building documentation using cargo");
+    let cargo = env::var("CARGO").unwrap_or_else(|_| {
+        println!("error: `cargo-deadlinks` must be run as either `cargo deadlinks` or with the `--dir` flag");
+        process::exit(1);
+    });
+    // Stolen from https://docs.rs/cargo_metadata/0.12.0/cargo_metadata/#examples
+    let mut cargo_process = Command::new(cargo)
+        .args(&["doc", "--no-deps", "--message-format", "json"])
+        .stdout(process::Stdio::piped())
+        // spawn instead of output() allows running deadlinks and cargo in parallel;
+        // this is helpful when you have many dependencies that take a while to document
+        .spawn()
+        .unwrap();
+    let reader = BufReader::new(cargo_process.stdout.take().unwrap());
+    // Originally written with jq:
+    // `select(.reason == "compiler-artifact") | .filenames[] | select(endswith("/index.html")) | rtrimstr("/index.html")`
+    let directories = Message::parse_stream(reader)
+        .filter_map(|message| match message.unwrap() {
```

Comment:
You might want to just filter out messages that failed to parse instead of panicking.